### PR TITLE
Remove Lucene split packages

### LINKED ISSUE #81981: Fix split package LazySoftDeletesDirectoryReaderWrapper
The last split package issue that we have between server and lucene is LazySoftDeletesDirectoryReaderWrapper. This was added early in 2021, so not one of the legacy hangovers from years gone by.

By inspecting the code, the three package-private member dependencies in `org.apache.lucene.index` become clear:
1. access to field SegmentReader.isNRT, 
2. CacheKey::new, and 
3. PendingSoftDeletes::applySoftDeletes.

First, LazySoftDeletesDirectoryReaderWrapper is only useful for read-only indices that are lazily loaded (a.k.a. frozen tier), so it  is doubtful that Lucene would find it generally useful.

A note on each of the specific package-private API points:
1. Access to field SegmentReader.isNRT is only for assertion code, and exposing an accessor from Lucene for this seems like it could be reasonable.
2. The CacheKey constructor is deliberately non-accessible from outside the package. There were past issues where problems arose with short-lived reader wrappers that would not reuse caches across queries. Maybe there's a way to make DelegatingCacheHelper only work with a DirectoryReader, since directory readers would generally not be short-lived. Or otherwise exposing DelegatingCacheHelper as a more generally useful API in Lucene itself.
3. For PendingSoftDeletes::applySoftDeletes one possibility is to just duplicate the code in ES. There are additional package-private issues when this is attempted, but maybe they are more straightforward to resolve.

[ the above notes are a summary of a discussion that included @romseygeek @jpountz, and @ywelsch ]

### PR DESCRIPTION
**Pull request by grcevski**
This PR proposes to fix the Lucene split package issue in LazySoftDeletesDirectoryReaderWrapper.

I did two things to make this work, one is a hack until we perhaps make a change in Lucene:

- (hack) I use SoftDeletesDirectoryReaderWrapper to make me a fresh CacheKey since we can't call the constructor.
- I copy pasted the applySoftDeletes code and removed the bits that were irrelevant (and causing us to bring more packages). Fortunately, the call to DocValuesFieldExistsQuery.getDocValuesDocIdSetIterator(field, reader) can't get us a reader back that has a non-null hasValue, so we could live without the iterator in DocValuesFieldUpdates (which is package private).

If we manage to get the DelegatingCacheHelper implementation as a public implementation in Lucene, then we get much better implementation.

closes #81981

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/engine/LazySoftDeletesDirectoryReaderWrapper.java [hunk 1]
```diff
@@ -260,4 +276,26 @@ public boolean initialized() {
             return materializedBits != null;
         }
     }
+
+    private static CacheKey getCacheKey(DirectoryReader reader, String field) throws IOException {
+        SoftDeletesDirectoryReaderWrapper wrapper = new SoftDeletesDirectoryReaderWrapper(reader, field);
```
Comment (grcevski) [inline server/src/main/java/org/elasticsearch/index/engine/LazySoftDeletesDirectoryReaderWrapper.java:281]: Hm, I realized there's an issue if reader.getReaderCacheHelper() is null. I can easily work around the null pointer here, but I need to somehow get a key inside the leaf and codec readers.

File: server/src/main/java/org/elasticsearch/index/engine/LazySoftDeletesDirectoryReaderWrapper.java [hunk 2]
```diff
@@ -260,4 +281,32 @@ public boolean initialized() {
...
+                // now that we know we deleted it and we fully control the hard deletes we can do correct
+                // accounting
+                // below.
+            }
+        }
+        return newDeletes;
+    }
+
+    static CacheKey createCacheKey(DirectoryReader in) {
+        try (DirectoryReader clone = DirectoryReader.open(in.directory())) {
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/index/engine/LazySoftDeletesDirectoryReaderWrapper.java:303]: This feels like too much of a hack to bring into ES. It's effectively opening up the index multiple times now whenever a search hits a shard in the frozen tier (one extra time per lucene index, and then once per segment in the Lucene index) . Let's rather work towards exposing the right abstraction in Lucene.

Comment (grcevski) [reply server/src/main/java/org/elasticsearch/index/engine/LazySoftDeletesDirectoryReaderWrapper.java:303]: Yeah I agree, it's a total hack. I couldn't figure out a better way to get to a unique CacheKey. Let's try to get a change in Lucene.

**Comment by grcevski** [comment]
@elasticmachine run Check labels

**Review by ChrisHegarty — APPROVED** [review]
LGTM

**Pull request merged**